pkg/certs.go

```diff
@@ -1,12 +1,86 @@
 package courier
 
 import (
+	"encoding/base64"
+	"errors"
 	"net/http"
 
 	"github.com/gin-gonic/gin"
 	"github.com/trisacrypto/courier/pkg/api/v1"
+	"github.com/trisacrypto/courier/pkg/store"
+	"github.com/trisacrypto/trisa/pkg/trust"
 )
 
+// StoreCertificate decodes a base64-encoded certificate in the request, decrypts it
+// using the password in the store, and stores the decrypted certificate in the store.
+// The NoDecrypt option can be used to skip the decryption and store the certificate in
+// its encrypted form.
+func (s *Server) StoreCertificate(c *gin.Context) {
+	var (
+		err error
+		req *api.StoreCertificateRequest
+	)
+
+	id := c.Param("id")
+	ctx := c.Request.Context()
+
+	// Parse the request body
+	req = &api.StoreCertificateRequest{}
+	if err := c.BindJSON(req); err != nil {
+		c.JSON(http.StatusBadRequest, api.ErrorResponse(err))
+		return
+	}
+
+	// Certificate is required
+	if req.Base64Certificate == "" {
+		c.JSON(http.StatusBadRequest, api.ErrorResponse("missing certificate in request"))
+		return
+	}
+
+	// Decode the certificate data from the request
+	var data []byte
+	if data, err = base64.StdEncoding.DecodeString(req.Base64Certificate); err != nil {
+		c.JSON(http.StatusBadRequest, api.ErrorResponse(err))
+		return
+	}
+
+	if !req.NoDecrypt {
+		// If decryption is enabled, retrieve the pkcs12 password from the store
+		var password []byte
+		if password, err = s.store.GetPassword(ctx, id); err != nil {
+			if errors.Is(err, store.ErrNotFound) {
+				c.JSON(http.StatusNotFound, api.ErrorResponse("pkcs12 password not found, unable to decrypt certificate"))
+				return
+			}
+
+			c.JSON(http.StatusInternalServerError, api.ErrorResponse(err))
+			return
+		}
+
+		// Decrypt the certificate using the password
+		var provider *trust.Provider
+		if provider, err = trust.Decrypt(data, string(password)); err != nil {
+			c.JSON(http.StatusConflict, api.ErrorResponse("failed to decrypt certificate with stored pkcs12 password"))
```

Comment:
I'm not sure what status code to return here. I didn't think 400 made sense because at this point we have already determined that the request is valid; the error here would indicate that either the stored password is wrong or the certificate provided in the request was wrong; or the ID in the request does not match the certificate.

Reply:
I was thinking maybe 422?  Based on the description, it appears to suggest that the request was formatted correctly but there was an issue processing data that was passed, which means that the caller needs to inspect the data that was passed.  

Here is a description of the status code:

```
422: Unprocessable Entity

A 422 Unprocessable Entity error status code indicates an issue with the contents of the syntax of a request. The arrangement of the request was understood by the server, but the fields within the request are invalid or does not match what the server expects. Like the 102 Processing and 207 Multi-Status status codes, a 422 Unprocessable Entity error code is part of the WebDAV protocol and often used with web services/APIs. Generally, a 400 Bad Request is the recommend response, but if WebDAV is supported, then the 422 Unprocessable Entity should be used. See [RFC4918, Section 11.2](https://tools.ietf.org/html/rfc4918#section-11.2) for more information.
```

Reply:
Based on the description it sounds like that's specific to the WebDAV protocol, and in this case the request is actually valid, it's just the state on the server that's probably wrong or the password was wrong, so I'm leaning towards 409.